### DISCUSSION TITLE
bsp/

### DISCUSSION THREAD

**Discussion by giscus** [radxa-repo/bsp#110] [Documents Comment]
# bsp/

https://radxa-repo.github.io/bsp/

**Comment by cichy977** [comment]
This example is not working properly, there is different list of parameters for ./bsp, log behind:

./bsp radxa-zero
'docker' backend is selected, but the functionality is actually provided by 'podman' backend. Updating accordingly...
WARN[0000] "/" is not a shared mount, this could cause issues or missing mounts with rootless containers
Trying to pull ghcr.io/radxa-repo/bsp:main...
Getting image source signatures
Copying blob 34086a0c1cc9 skipped: already exists
Copying blob 0a7a0c2c50e9 skipped: already exists
Copying blob 29f873e2e3f8 skipped: already exists
Copying blob 1874e107f449 skipped: already exists
Copying config 5e2d53934c done
Writing manifest to image destination
Storing signatures
5e2d53934c4d9706d3ae5baeb1d9aa96356f8ceb163a1c1f5d33a540d7b09b7d
'radxa-zero' is not part of the supported components: linux u-boot
Assume it is a product.
'radxa-zero' is supported by the following linux profiles: stable
'radxa-zero' is supported by the following u-boot profiles: latest
Do you want to build supported components with the first profile listed? [Y/n]

Could you comment, please?

**Reply by RadxaYuntian** [reply]
What's the issue here? It asks you if you want to build them, and if you press enter it should build them?

**Reply by cichy977** [reply]
Yes, but I have the issue described in next comment.

**Comment by cichy977** [comment]
There is another issue with fixdep script, log behind:

Patchset 0200-upstream has been applied.
/home/user/bsp
Initialize .config with defconfig
make: Entering directory '/home/cichuta/bsp/.src/linux'
  HOSTCC  scripts/basic/fixdep
/bin/sh: 1: scripts/basic/fixdep: Exec format error
make[1]: *** [scripts/Makefile.host:111: scripts/basic/fixdep] Error 126
make[1]: *** Deleting file 'scripts/basic/fixdep'
make: *** [Makefile:640: scripts_basic] Error 2
make: Leaving directory '/home/user/bsp/.src/linux'
f31d3af53a2d2a1c809ba2363e94470c8989bc02812806b92c8ba1fcdde0e44d

**Reply by RadxaYuntian** [reply]
You didn't follow the guide where it tells you to run `sudo apt install -y git qemu-user-static binfmt-support`.

**Reply by cichy977** [reply]
Yes, I did it.

**Reply by cichy977** [reply]
sudo apt install -y git qemu-user-static binfmt-support
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
binfmt-support is already the newest version (2.2.1-2).
git is already the newest version (1:2.34.1-1ubuntu1.11).
qemu-user-static is already the newest version (1:6.2+dfsg-2ubuntu6.22).
The following packages were automatically installed and are no longer required:
  linux-headers-5.15.0-88 linux-headers-5.15.0-88-generic linux-tools-5.15.0-88 linux-tools-5.15.0-88-generic
Use 'sudo apt autoremove' to remove them.
0 upgraded, 0 newly installed, 0 to remove and 211 not upgraded.

**Reply by cichy977** [reply]
There is something wrong witch HOSTCC:

./bsp radxa-cm5-io
'docker' backend is selected, but the functionality is actually provided by 'podman' backend. Updating accordingly...
Trying to pull ghcr.io/radxa-repo/bsp:main...
Getting image source signatures
Copying blob 34086a0c1cc9 skipped: already exists
Copying blob 1874e107f449 skipped: already exists
Copying blob 0a7a0c2c50e9 skipped: already exists
Copying blob 29f873e2e3f8 skipped: already exists
Copying config 5e2d53934c done
Writing manifest to image destination
Storing signatures
5e2d53934c4d9706d3ae5baeb1d9aa96356f8ceb163a1c1f5d33a540d7b09b7d
'radxa-cm5-io' is not part of the supported components: linux u-boot
Assume it is a product.
'radxa-cm5-io' is supported by the following linux profiles: rk2312 rockchip
'radxa-cm5-io' is supported by the following u-boot profiles: rknext
Do you want to build supported components with the first profile listed? [Y/n] Y
/home/cichuta/bsp/.src/linux /home/cichuta/bsp
hint: Using 'master' as the name for the initial branch. This default branch name
hint: is subject to change. To configure the initial branch name to use in all
hint: of your new repositories, which will suppress this warning, call:
hint:
hint:   git config --global init.defaultBranch <name>
hint:
hint: Names commonly chosen instead of 'master' are 'main', 'trunk' and
hint: 'development'. The just-created branch can be renamed via this command:
hint:
hint:   git branch -m <name>
Initialized empty Git repository in /home/cichuta/bsp/.src/linux/.git/
remote: Enumerating objects: 87543, done.
remote: Counting objects: 100% (87543/87543), done.
remote: Compressing objects: 100% (79316/79316), done.
remote: Total 87543 (delta 9237), reused 36062 (delta 7337), pack-reused 0 (from 0)
Receiving objects: 100% (87543/87543), 244.28 MiB | 9.54 MiB/s, done.
Resolving deltas: 100% (9237/9237), done.
From https://github.com/radxa/kernel
 * branch                linux-6.1-stan-rkr1 -> FETCH_HEAD
 * [new branch]          linux-6.1-stan-rkr1 -> 8f2d0659fd3e83cfb9e4231704fdc9b1fb4acd0b/linux-6.1-stan-rkr1
Updating files: 100% (82669/82669), done.
HEAD is now at 1b2f1f909 arm64: dts: rockchip: add medge-rk3528a io board
HEAD is now at 1b2f1f909 arm64: dts: rockchip: add medge-rk3528a io board
Running custom_source_action from /home/cichuta/bsp/linux/rk2312/0001-common/overlays.sh
/home/cichuta/bsp/linux/rk2312/0001-common /home/cichuta/bsp/.src/linux /home/cichuta/bsp
/home/cichuta/bsp/.src/radxa-overlays /home/cichuta/bsp/linux/rk2312/0001-common /home/cichuta/bsp/.src/linux /home/cichuta/bsp
hint: Using 'master' as the name for the initial branch. This default branch name
hint: is subject to change. To configure the initial branch name to use in all
hint: of your new repositories, which will suppress this warning, call:
hint:
hint:   git config --global init.defaultBranch <name>
hint:
hint: Names commonly chosen instead of 'master' are 'main', 'trunk' and
hint: 'development'. The just-created branch can be renamed via this command:
hint:
hint:   git branch -m <name>
Initialized empty Git repository in /home/cichuta/bsp/.src/radxa-overlays/.git/
fatal: ambiguous argument 'FETCH_HEAD': unknown revision or path not in the working tree.
Use '--' to separate paths from revisions, like this:
'git <command> [<revision>...] -- [<file>...]'
remote: Enumerating objects: 613, done.
remote: Counting objects: 100% (613/613), done.
remote: Compressing objects: 100% (351/351), done.
remote: Total 613 (delta 425), reused 341 (delta 249), pack-reused 0 (from 0)
Receiving objects: 100% (613/613), 175.98 KiB | 2.12 MiB/s, done.
Resolving deltas: 100% (425/425), done.
From https://github.com/radxa-pkg/radxa-overlays
 * branch            04de4cab2c5eba53cb2caff940b78822dac50ab5 -> FETCH_HEAD
HEAD is now at 04de4ca Merge pull request #365 from Ken-Vamrs/main
HEAD is now at 04de4ca Merge pull request #365 from Ken-Vamrs/main
/home/cichuta/bsp/linux/rk2312/0001-common /home/cichuta/bsp/.src/linux /home/cichuta/bsp
/home/cichuta/bsp/.src/linux /home/cichuta/bsp
Applying: Disable warnings
Checking patch Makefile...
Hunk #1 succeeded at 1083 (offset -5 lines).
Applied patch Makefile cleanly.
Applying: fix: import namespace DMA_BUF before using it
Checking patch drivers/usb/gadget/function/uvc_queue.c...
Applied patch drivers/usb/gadget/function/uvc_queue.c cleanly.
Patchset 0100-vendor has been applied.
/home/cichuta/bsp
Initialize .config with rockchip_linux_defconfig
make: Entering directory '/home/cichuta/bsp/.src/linux'
  HOSTCC  scripts/basic/fixdep
/bin/sh: 1: scripts/basic/fixdep: Exec format error
make[1]: *** [scripts/Makefile.host:111: scripts/basic/fixdep] Error 126
make[1]: *** Deleting file 'scripts/basic/fixdep'
make: *** [Makefile:640: scripts_basic] Error 2
make: Leaving directory '/home/cichuta/bsp/.src/linux'
61b007e9fa9ed180d332e7e70c067d164b8842738a54dbe6aa8b98fe2ca764b6

For my opinion for Makefile.host should be use CC (real host gcc, not cross like it is now in your build script), now I have x86_64 PC but build script used cross compile for build scripts/basic/fixdep and next try to run on x86 machine, so the error occured:
/bin/sh: 1: scripts/basic/fixdep: Exec format error

**Reply by RadxaYuntian** [reply]
No, host tools MUST be cross built to arm64, since those will be copied to the kernel header package, and shipping x64 tools in arm64 package is not acceptable. This is a [long known issue](https://www.usenix.org/sites/default/files/conference/protected-files/srecon19apac_slides_korchagin.pdf).

You will need to check why your system cannot have binfmt working. Maybe fully upgrade your system since there are 211 packages that can be updated, and reboot. We use [the same packages on GitHub hosted Ubuntu runner](https://github.com/radxa-repo/bsp/blob/main/action.yaml#L30) and they work fine.

**Reply by cichy977** [reply]
I am using Ubuntu 22.04 from Windows 10 WSL2 system.

**Reply by cichy977** [reply]
And I tested build the downloaded kernel source to .src/ "by hand":
 
make ARCH=arm64 CROSS_COMPILE=aarch64-linux-gnu- rockchip_linux_defconfig
  HOSTCC  scripts/basic/fixdep
  HOSTCC  scripts/kconfig/conf.o
  HOSTCC  scripts/kconfig/confdata.o
  HOSTCC  scripts/kconfig/expr.o
  LEX     scripts/kconfig/lexer.lex.c
  YACC    scripts/kconfig/parser.tab.[ch]
  HOSTCC  scripts/kconfig/lexer.lex.o
  HOSTCC  scripts/kconfig/parser.tab.o
  HOSTCC  scripts/kconfig/preprocess.o
  HOSTCC  scripts/kconfig/symbol.o
  HOSTCC  scripts/kconfig/util.o
  HOSTLD  scripts/kconfig/conf
boolean symbol CEC_CORE tested for 'm'? test forced to 'n'
#
# configuration written to .config
#
make ARCH=arm64 CROSS_COMPILE=aarch64-linux-gnu- -j8
(build output ...)

there is no problem with fixdep

**Reply by cichy977** [reply]
I can provide any log that you need to investigate this issue.

**Reply by RadxaYuntian** [reply]
Please use `file` to check the architecture of the generated fixdep.

**Reply by RadxaYuntian** [reply]
If you can confirm the issue is due to WSL environment (so a regular VM should work fine), then you can create an issue for it.

**Reply by Neirop** [reply]
I get the same error, I'm using the bsp docker image:

root@c8ff94db748e:/bsp# ./bsp -n radxa-zero3
...
Patchset 0100-vendor has been applied.
/bsp
Initialize .config with rockchip_linux_defconfig
make: Entering directory '/bsp/.src/linux'
  HOSTCC  scripts/basic/fixdep
/bin/sh: 1: scripts/basic/fixdep: Exec format error
make[1]: *** [scripts/Makefile.host:95: scripts/basic/fixdep] Error 126
make[1]: *** Deleting file 'scripts/basic/fixdep'
make: *** [Makefile:563: scripts_basic] Error 2
make: Leaving directory '/bsp/.src/linux'

root@c8ff94db748e:/bsp# file .src/u-boot/scripts/basic/fixdep
.src/u-boot/scripts/basic/fixdep: ELF 64-bit LSB pie executable, x86-64, version 1 (SYSV), dynamically linked, interpreter /lib64/ld-linux-x86-64.so.2, BuildID[sha1]=9de3b593809e2937ffba95800fdc1b4af7d86f19, for GNU/Linux 3.2.0, not stripped

**Reply by RadxaYuntian** [reply]
Have you tried this? https://github.com/radxa-repo/bsp/discussions/110#discussioncomment-10555059

**Comment by avri-altman-sndk** [comment]
hi, 
I noticed that you are providing a ROCK 5B Debian (kernel 6.1) image - https://docs.radxa.com/en/rock5/rock5b/download

However, I followed the instructions in https://docs.radxa.com/en/rock5/rock5b/low-level-dev/kernel and downloaded the bsp code:
avri@avri-office:~/src/sandisk/rock5B/bsp$ git remote -vv
origin  https://github.com/radxa-repo/bsp.git (fetch)
origin  https://github.com/radxa-repo/bsp.git (push)
avri@avri-office:~/src/sandisk/rock5B/bsp$

But the kernel version you provides is 5.10 :
avri@avri-office:~/src/sandisk/rock5B/bsp/.src/linux$ head Makefile
# SPDX-License-Identifier: GPL-2.0
VERSION = 5
PATCHLEVEL = 10
SUBLEVEL = 110
EXTRAVERSION =
NAME = Dare mighty things

I wasn't able to find in your repository any branch that contains a more advanced kernel.
Can you refer me to a branch that contains a more recent kernel?
Thanks,
Avri

**Reply by avri-altman-sndk** [reply]
I ran git fetch and I see the following 6.1 branches:
remotes/8f2d0659fd3e83cfb9e4231704fdc9b1fb4acd0b/kernel-6.1-2024_01_02              5baeec1f9e6c arm64: dts: rockchip: add rock-2a
  remotes/8f2d0659fd3e83cfb9e4231704fdc9b1fb4acd0b/linux-6.1-stan-rkr1                3e26818dcaaa arm64: dts: rockchip: modify fan pwm period to 60us
  remotes/8f2d0659fd3e83cfb9e4231704fdc9b1fb4acd0b/linux-6.1-stan-rkr4.1              3bf163742624 drm/sched: Use drm sched lockdep map for submit_wq
  remotes/8f2d0659fd3e83cfb9e4231704fdc9b1fb4acd0b/linux-6.1-stan-rkr4.1-buildroot    a1a382b215fc arm64: dts: rockchip: add radxa rock2 boards

I tried to compile few of them with the --no-prepare-source option - none of them compiled.
Thanks,
Avri

**Comment by houjw-hx** [comment]
Which parameter is set for radxa-zero3w？Should I continue it?
~/bsp$ ./bsp radxa-zero3w
main: Pulling from radxa-repo/bsp
Digest: sha256:01a853cd42c4dd56ef8a38039d581bc2655d9710bb714b6d26263c16af865caf
Status: Image is up to date for ghcr.io/radxa-repo/bsp:main
ghcr.io/radxa-repo/bsp:main
'radxa-zero3w' is not part of the supported components: linux u-boot
Assume it is a product.
'radxa-zero3w' is not supported by linux.
'radxa-zero3w' is not supported by u-boot.
Do you want to build supported components with the first profile listed? [Y/n]

**Reply by RadxaYuntian** [reply]
It is just `radxa-zero3`.

**Reply by houjw-hx** [reply]
Actually I tried radxa-zero3 firstly and it is not supported by linux. I want to change another image sensor, so need modify the media driver.

~/bsp$ ./bsp radxa-zero3
main: Pulling from radxa-repo/bsp
Digest: sha256:01a853cd42c4dd56ef8a38039d581bc2655d9710bb714b6d26263c16af865caf
Status: Image is up to date for ghcr.io/radxa-repo/bsp:main
ghcr.io/radxa-repo/bsp:main
'radxa-zero3' is not part of the supported components: linux u-boot
Assume it is a product.
'radxa-zero3' is not supported by linux.
'radxa-zero3' is supported by the following u-boot profiles: rk2410
Do you want to build supported components with the first profile listed? [Y/n]

**Reply by RadxaYuntian** [reply]
You will need to revert [this commit](https://github.com/radxa-repo/bsp/commit/e954b75739ea20a1a994ad8c20e41245bff57344), as we are in the process of updating ZERO3 to Rockchip 6.1 kernel. As such, old 5.10 kernel has removed its support in case if we need to create another kernel release for this code branch. This way we won't have a 5.10 kernel releasing untested for ZERO 3.

**Comment by dhruv77777** [comment]
Facing this error:

dpkg: error processing archive u-boot-rk2410_2017.09-1_arm64.deb (--install): conflicting packages - not installing u-boot-rk2410 Errors were encountered while processing: u-boot-rk2410_2017.09-1_arm64.deb sudo dpkg -i u-boot-radxa-cm5-io_2017.09-1_all.deb (Reading database ... 65922 files and directories currently installed.)dxa-cm5-io_2017.09-1_all.deb Preparing to unpack u-boot-radxa-cm5-io_2017.09-1_all.deb ... Unpacking u-boot-radxa-cm5-io (2017.09-1) over (2017.09-1) ... dpkg: dependency problems prevent configuration of u-boot-radxa-cm5-io: u-boot-radxa-cm5-io depends on u-boot-rk2410; however: Package u-boot-rk2410 is not installed. dpkg: error processing package u-boot-radxa-cm5-io (--install): dependency problems - leaving unconfigured Errors were encountered while processing: u-boot-radxa-cm5-io radxa@radxa-cm5-io:/usr/lib/u-boot/radxa-cm5-io$

**Reply by RadxaYuntian** [reply]
You installed the metapackage, not the binary package. Please check [here](https://radxa-repo.github.io/bsp/artifacts.html) for more.
